### PR TITLE
Implement the Runtime Cache

### LINKED ISSUE #680: Implement the Runtime Cache
### Summary
Now that we have a concrete design for the runtime cache as a result of #679, that cache needs to be implemented.

### Value
We will have the implementation of the cache ready to be connected up to the execution service.

### Specification

- [ ] Implement the specification designed in #679.
- [ ] Implement a comprehensive benchmark suite for the cache.
- [ ] Implement comprehensive tests for the cache, including property-based tests.

### Acceptance Criteria & Test Cases

- The cache has been implemented and meets its specification.
- It is accompanied by a regression suite of benchmarks and tests.

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #680

Changelog:

- add: Implement calculation of _TextEdit_ dependencies using the _DataflowAnalysis_
- add: `RuntimeCache` test & bench
- update: Correctly handle `InvalidatedExpressions` parameter of the _recompute_ request
- update: Store parsed `IR` in `Module`
- refactor: rename `Module#getScope` -> `Module#parseScope`

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/docs/style-guide/scala.md) and [Java](https://github.com/luna/enso/blob/master/docs/style-guide/java.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

**Comment by kustosz** [comment]
One more thing: It seems to me that with the current implement, if an ID disappears from the metadata, the cached value will remain in the caches forever.

File: engine/runtime/src/main/scala/org/enso/compiler/context/DiffChangeset.scala [hunk 1]
```diff
@@ -1,41 +1,60 @@
...
+
+  /**
+    * Traverses the IR and returns a list of all IR nodes affected by the edit
+    * using the [[DataflowAnalysis]] information.
+    *
+    * @param edit the text edit.
+    * @return the list of all IR nodes affected by the edit.
+    */
+  def compute(edit: TextEdit): Seq[IR.ExternalId] = {
+    ir.getMetadata(DataflowAnalysis)
```
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/compiler/context/DiffChangeset.scala:26]: This shouldn't return an empty sequence if the dependency analysis metadata is missing. It's indicative of a programmer bug that said metadata doesn't exist at this point, so we should crash loudly with a `CompilerError`.

File: engine/runtime/src/main/scala/org/enso/compiler/context/DiffChangeset.scala [hunk 2]
```diff
@@ -1,41 +1,60 @@
...
 import scala.collection.mutable
 
-final class DiffChangeset {
+/**
+  * Compute invalidated expressions.
+  *
+  * @param source the text source.
+  * @param ir the IR node.
+  */
+final class DiffChangeset(source: CharSequence, ir: IR) {
```
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/compiler/context/DiffChangeset.scala:16]: A changeset is implicitly for a diff, so this name is redundant. Just call it `Changeset`.

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/ExecutionContextState.scala [hunk 3]
```diff
@@ -21,6 +21,20 @@ object ExecutionContextState {
...
+    ExecutionContextState(mutable.Stack.empty, VisualisationHolder.empty)
+}
+
+/**
+  * Stack frame of the context.
+  *
+  * @param item the stack item.
+  * @param cache the cache of this stack frame.
+  */
+case class StackFrame(item: StackItem, cache: RuntimeCache)
```
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/interpreter/instrument/ExecutionContextState.scala:34]: I don't like calling this `StackFrame`. It implies that it's a raw stack frame from the interpreter which it isn't.

File: engine/runtime/src/main/java/org/enso/interpreter/instrument/RuntimeCache.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+  private final Map<UUID, SoftReference<Object>> cache = new HashMap<>();
+
+  /** Add computed value to the cache. */
+  public Object put(UUID key, Object value) {
+    return cache.put(key, new SoftReference<>(value));
+  }
+
+  /** Get value from the cache. */
+  public Object get(UUID key) {
```
[the remark this answers is not in the document]
Comment (4e6) [reply engine/runtime/src/main/java/org/enso/interpreter/instrument/RuntimeCache.java:19]: The cache is used in the Instrument during the program execution. I thought we don't want to create intermediate objects there.

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/ExecutionContextState.scala (hunk 3 above)
Comment (4e6) [reply engine/runtime/src/main/scala/org/enso/interpreter/instrument/ExecutionContextState.scala]: Ugh, it's hard to find an alternative name for аn established concept. I just called it `Frame`

Comment (kustosz) [reply engine/runtime/src/main/scala/org/enso/interpreter/instrument/ExecutionContextState.scala]: ^ That's actually worse... `Frame` is an actual name clash, not just conceptual one. I'd stick to `StackFrame`

File: engine/runtime/src/main/scala/org/enso/compiler/context/Changeset.scala [hunk 5]
```diff
@@ -1,41 +1,62 @@
...
+    *
+    * @param edit the text edit.
+    * @throws CompilerError if the IR is missing DataflowAnalysis metadata
+    * @return the list of all IR nodes affected by the edit.
+    */
+  @throws[CompilerError]
+  def compute(edit: TextEdit): Seq[IR.ExternalId] = {
+    val metadata = ir
+      .getMetadata(DataflowAnalysis)
+      .getOrElse(throw new CompilerError("Empty DataflowAnalysis metadata"))
```
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/compiler/context/Changeset.scala:31]:
```suggestion
      .getOrElse(throw new CompilerError("Empty dataflow analysis metadata."))
```

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/ExecutionContextState.scala [hunk 6]
```diff
@@ -21,6 +21,20 @@ object ExecutionContextState {
...
+    ExecutionContextState(mutable.Stack.empty, VisualisationHolder.empty)
+}
+
+/**
+  * Stack frame of the context.
+  *
+  * @param item the stack item.
+  * @param cache the cache of this stack frame.
+  */
+case class Frame(item: StackItem, cache: RuntimeCache)
```
Comment (iamrecursion) [inline engine/runtime/src/main/scala/org/enso/interpreter/instrument/ExecutionContextState.scala:34]: Okay `Frame` is definitely worse than `StackFrame`. Try `InstrumentFrame`.

**Pull request merged**